src/lib/pubkey/pcurves/pcurves_impl.h

```diff
@@ -0,0 +1,816 @@
+
+      //static constexpr std::optional<Self> deserialize(std::span<const uint8_t> bytes) {}
+
+      constexpr const FieldElement& x() const { return m_x; }
+
+      constexpr const FieldElement& y() const { return m_y; }
+
+   private:
+      FieldElement m_x;
+      FieldElement m_y;
+};
+
+template <typename FieldElement, StringLiteral AS>
+class ProjectiveCurvePoint {
+   public:
+      // We can't pass a FieldElement directly because FieldElement is
+      // not "structural" due to having private members, so instead
+      // recreate it here from the string.
+      static const constexpr FieldElement A = FieldElement::constant(AS);
+
+      static const constinit bool A_is_minus_3 = (A == FieldElement::constant(-3));
+      static const constinit bool A_is_zero = (A == FieldElement::zero());
+
+      typedef ProjectiveCurvePoint<FieldElement, AS> Self;
+      typedef AffineCurvePoint<FieldElement> AffinePoint;
+
+      static constexpr Self from_affine(const AffinePoint& pt) { return ProjectiveCurvePoint(pt.x(), pt.y()); }
+
+      static constexpr Self identity() {
+         return Self(FieldElement::zero(), FieldElement::zero(), FieldElement::zero());
+      }
+
+      constexpr ProjectiveCurvePoint() :
+            m_x(FieldElement::zero()), m_y(FieldElement::zero()), m_z(FieldElement::zero()) {}
+
+      constexpr ProjectiveCurvePoint(const FieldElement& x, const FieldElement& y) :
+            m_x(x), m_y(y), m_z(FieldElement::one()) {}
+
+      constexpr ProjectiveCurvePoint(const FieldElement& x, const FieldElement& y, const FieldElement& z) :
+            m_x(x), m_y(y), m_z(z) {}
+
+      ProjectiveCurvePoint(const Self& other) = default;
+      ProjectiveCurvePoint(Self&& other) = default;
+      ProjectiveCurvePoint& operator=(const Self& other) = default;
+      ProjectiveCurvePoint& operator=(Self&& other) = default;
+
+      friend constexpr Self operator+(const Self& a, const Self& b) { return Self::add(a, b); }
+
+      friend constexpr Self operator+(const Self& a, const AffinePoint& b) { return Self::add_mixed(a, b); }
+
+      friend constexpr Self operator+(const AffinePoint& a, const Self& b) { return Self::add_mixed(b, a); }
+
+      constexpr Self& operator+=(const Self& other) {
+         (*this) = (*this) + other;
+         return (*this);
+      }
+
+      constexpr Self& operator+=(const AffinePoint& other) {
+         (*this) = (*this) + other;
+         return (*this);
+      }
+
+      friend constexpr Self operator-(const Self& a, const Self& b) { return a + b.negate(); }
+
+      constexpr bool is_identity() const { return z().is_zero(); }
+
+      template <typename Pt>
+      constexpr void conditional_add(bool cond, const Pt& pt) {
+         conditional_assign(cond, *this + pt);
+      }
+
+      void conditional_assign(bool cond, const Self& pt) {
+         m_x.conditional_assign(cond, pt.x());
+         m_y.conditional_assign(cond, pt.y());
+         m_z.conditional_assign(cond, pt.z());
+      }
```

Comment:
I'm assuming this is needed for a constant-time implementation? So far, I saw it as an unspoken rule that all methods that are "supposed to be constant-time" are prefixed with `ct_...`. I found this a helpful convention, that might be applicable here as well.

Reply:
In this code I'm going for the opposite convention namely that everything is constant time by default and anything that is intentionally variable time has a `_vartime` suffix.

(There are some current exceptions eg in the point multiplication that will be fixed before merging.)